### PR TITLE
Support multi GPU training using torchrun

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary
- Support multi GPU training using torchrun(torch.elastic)
- This is for Geti, so `--gpus` flag is used when using OTX cli.

### How to test
You can test by adding below line at the begging of command line.
`torchrun --standalone --nnodes=1 --nproc_per_node=2 --no_python`

For example,
`torchrun --standalone --nnodes=1 --nproc_per_node=2 --no_python otx train ...`

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: otx/algorithms/common/tasks/training_base.py [hunk 1]
```diff
@@ -126,6 +128,20 @@ def __init__(self, task_config, task_environment: TaskEnvironment, output_path:
...
+        if self._is_multi_gpu_training():
+            self._setup_multigpu_training()
+
+    @staticmethod
+    def _is_multi_gpu_training():
+        return "LOCAL_RANK" in os.environ
+
+    @staticmethod
+    def _setup_multigpu_training():
+        if not dist.is_initialized():
```
Comment (JihwanEom) [inline otx/algorithms/common/tasks/training_base.py:142]: Not sure, how about
```suggestion
        if not dist.is_initialized() and torch.cuda.is_available():
```

File: otx/cli/utils/multi_gpu.py [hunk 2]
```diff
@@ -152,15 +153,22 @@ def is_available(self) -> bool:
...
         Args:
             output_path (str): output path where task output are saved.
             optimized_hyper_parameters (ConfigurableParameters or None): hyper parameters reflecting HPO result.
+
+        Returns:
+            str:
+                If output_path is None, make a temporary directory and return it.
         """
+        if output_path is None:
+            output_path = tempfile.mkdtemp(prefix="OTX-multi-gpu-")
```
Comment (JihwanEom) [inline otx/cli/utils/multi_gpu.py]: Is it will be not removed automatically via ContextManager?

Comment (eunwoosh) [reply otx/cli/utils/multi_gpu.py]: No. I will update it to be removed automatically. Thanks for pointing it!

**Review by jaegukhyun — DISMISSED** [review]
I don't have big objection for this PR, It looks ok if the tests are passed. 
However, current base task(actually all of task) contains too many attributes and methods. 
In the next refactoring I'll remove attributes and methods as much as I can. 
And I think we should carefully add new attributes and new methods to base task.
I think we may split multi-gpu setting from base task such as multi-gpu-mixin.

File: otx/algorithms/common/tasks/training_base.py (hunk 1 above)
Comment (eunwoosh) [reply otx/algorithms/common/tasks/training_base.py:142]: Nice catch! but I think checking it in `_is_multi_gpu_training()` is better. I'll update it.

Comment (eunwoosh) [reply otx/algorithms/common/tasks/training_base.py:142]: I fixed.

File: otx/cli/utils/multi_gpu.py (hunk 2 above)
Comment (eunwoosh) [reply otx/cli/utils/multi_gpu.py]: I fixed.

**Review by JihwanEom — COMMENTED** [review]
LGTM, but failed with updated test case. T_T could you resolve this?

File: otx/algorithms/common/tasks/training_base.py [hunk 3]
```diff
@@ -126,6 +125,27 @@ def __init__(self, task_config, task_environment: TaskEnvironment, output_path:
         # to override configuration at runtime
         self.override_configs = {}  # type: Dict[str, str]
 
+        if self._is_multi_gpu_training():
+            self._setup_multigpu_training()
+
+    @staticmethod
```
Comment (sungmanc) [inline otx/algorithms/common/tasks/training_base.py:131]: Why did you set this function as static? It seems that this function is only used with `self.` and this function is a private function.

**Review by sungmanc — DISMISSED** [review]
LGTM, but left minor comment

File: otx/algorithms/common/adapters/mmcv/tasks/stage.py [hunk 4]
```diff
@@ -180,12 +180,10 @@ def __init_device(self):
...
-            gpu_ids = os.environ.get("CUDA_VISIBLE_DEVICES")
-            logger.info(f"CUDA_VISIBLE_DEVICES = {gpu_ids}")
-            if gpu_ids is not None:
-                self.cfg.gpu_ids = range(len(gpu_ids.split(",")))
-            else:
-                self.cfg.gpu_ids = range(1)
+            self.cfg.gpu_ids = range(1)
+        else:
+            if len(self.cfg.gpu_ids) != 0:
+                raise ValueError("length of cfg.gpu_ids should be 1 when training with single GPU.")
```
Comment (goodsong81) [inline otx/algorithms/common/adapters/mmcv/tasks/stage.py]: elif?

And there is mismatch between the condition and error message. 0? 1?

File: otx/algorithms/common/tasks/training_base.py (hunk 3 above)
Comment (goodsong81) [reply otx/algorithms/common/tasks/training_base.py:131]: Maybe it's not using any internal attribute of `self`?
Static methods are not necessarily public.

File: otx/algorithms/common/tasks/training_base.py [hunk 5]
```diff
@@ -126,6 +125,27 @@ def __init__(self, task_config, task_environment: TaskEnvironment, output_path:
...
+    def _is_multi_gpu_training():
+        return "LOCAL_RANK" in os.environ and torch.cuda.is_available()
+
+    @staticmethod
+    def _setup_multigpu_training():
+        if not dist.is_initialized():
+            torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
+            dist.init_process_group(backend="nccl", init_method="env://", timeout=timedelta(seconds=30))
+            logger.info(f"dist info world_size = {dist.get_world_size()}, rank = {dist.get_rank()}")
```
Comment (goodsong81) [inline otx/algorithms/common/tasks/training_base.py]: f"Dist info: rank {n} / {N} world_size" might be more informatic.

File: otx/cli/tools/train.py [hunk 6]
```diff
@@ -224,6 +223,8 @@ def main():  # pylint: disable=too-many-branches
                 str(config_manager.output_path), hyper_parameters if args.enable_hpo else None
             )
 
+    task = task_class(task_environment=environment, output_path=str(config_manager.output_path / "logs"))
```
Comment (goodsong81) [inline otx/cli/tools/train.py:231]: path.join?

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Thank you for the great work. All look good to me except for a few things to fix.

File: otx/algorithms/common/tasks/training_base.py (hunk 3 above)
Comment (eunwoosh) [reply otx/algorithms/common/tasks/training_base.py:131]: Thanks for answer :)  As Songki said, I also think that static method don't have to be public method. And also being staticmethod doesn't have any side effect but have advantage even by not to need information of self.

File: otx/algorithms/common/adapters/mmcv/tasks/stage.py (hunk 4 above)
Comment (eunwoosh) [reply otx/algorithms/common/adapters/mmcv/tasks/stage.py]: you are right. thanks for pointing. I will fix it

File: otx/algorithms/common/tasks/training_base.py (hunk 5 above)
Comment (eunwoosh) [reply otx/algorithms/common/tasks/training_base.py]: I agree. I'll change it

File: otx/cli/tools/train.py (hunk 6 above)
Comment (eunwoosh) [reply otx/cli/tools/train.py:231]: `config_manager.output_path` is Path class, so honestly I think current implementation is more natural. Is it better to use `path.join`?

File: otx/cli/utils/hpo.py [hunk 7]
```diff
@@ -553,7 +554,11 @@ def run_hpo(
             "Currently supported task types are classification, detection, segmentation and anomaly"
             f"{task_type} is not supported yet."
         )
-        return None
+        return environment
+
+    if "TORCHELASTIC_RUN_ID" in os.environ:
+        logger.warning("OTX is train by torchrun. HPO isn't available.")
```
Comment (supersoob) [inline otx/cli/utils/hpo.py]:
```suggestion
        logger.warning("OTX is trained by torchrun. HPO isn't available.")
```

File: otx/cli/utils/multi_gpu.py [hunk 8]
```diff
@@ -284,6 +305,11 @@ def _spawn_multi_gpu_processes(self, output_path: str) -> List[mp.Process]:
             task_p.start()
             processes.append(task_p)
 
+        if origin_cuda_visible_devices is None:
+            del os.environ["CUDA_VISIBLE_DEVICES"]
+        else:
+            os.environ["CUDA_VISIBLE_DEVICES"] = origin_cuda_visible_devices
```
Comment (supersoob) [inline otx/cli/utils/multi_gpu.py:311]: I'm just curious, why do we need this?

**Review by supersoob — COMMENTED** [review]
Nice work:))

File: otx/cli/utils/multi_gpu.py (hunk 8 above)
Comment (eunwoosh) [reply otx/cli/utils/multi_gpu.py:311]: To apply `CUDA_VISIBLE_DEVICES` environment variable to child process, `CUDA_VISIBLE_DEVICES` should be changed before spawning child process in parent process. So, this code is for reverting `CUDA_VISIBLE_DEVICES` value as before `CUDA_VISIBLE_DEVICES` is modified.

**Comment by sungmanc** [comment]
Will we enable the CLI tests for multi-GPU training in the next PR?

**Comment by eunwoosh** [comment]
> Will we enable the CLI tests for multi-GPU training in the next PR?

Yes. I think that's an urgent issue. I'll re-enable multi GPU TC ASAP after this PR is merged.

**Review by sungmanc — APPROVED** [review]

File: otx/cli/tools/train.py (hunk 6 above)
Comment (goodsong81) [reply otx/cli/tools/train.py:231]: Nope.

**Review by goodsong81 — APPROVED** [review]

**Pull request merged**